05_Image/bedrock-titan-embed-image.ipynb

```diff
@@ -0,0 +1,1226 @@
+    "    title = f\"{embed_dir}/{title}.png\"\n",
+    "    titles.append(title)\n",
+    "    images[0].save(title, format=\"png\")\n",
+    "    print(f\"generated {title}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "900115e4-5be3-40c4-9982-b579c3f3f863",
+   "metadata": {},
+   "source": [
+    "# Multimodal Dataset Indexing"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bcac42d5-810e-4ad4-bd09-5ed89991e9d9",
+   "metadata": {},
+   "source": [
+    "The following function converts image, and optionally, text, into multimodal embeddings."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "801aa752-afe0-47bc-b73a-ec2667c9559a",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "def titan_multimodal_embedding(\n",
+    "    image_path:str=None,  # maximum 2048 x 2048 pixels\n",
+    "    description:str=None, # English only and max input tokens 128\n",
+    "    dimension:int=1024,   # 1,024 (default), 384, 256\n",
+    "    model_id:str=\"amazon.titan-embed-image-v1\"\n",
+    "):\n",
+    "\n",
+    "    embedding_config = {\n",
+    "        \"embeddingConfig\": { \n",
+    "             \"outputEmbeddingLength\": dimension\n",
+    "         }\n",
+    "    }\n",
+    "\n",
+    "    # You can specify either text or image or both\n",
+    "    if image_path:\n",
+    "        with open(image_path, \"rb\") as image_file:\n",
+    "            input_image = base64.b64encode(image_file.read()).decode('utf8')\n",
+    "        if description:\n",
+    "            print(\"text+image\")\n",
+    "            body = {\n",
+    "                    \"inputText\": description,\n",
+    "                    \"inputImage\": input_image\n",
+    "                    }\n",
+    "            \n",
+    "        else:\n",
+    "            print(\"image\")\n",
+    "            body = {\n",
+    "                    \"inputImage\": input_image\n",
+    "                    }\n",
+    "    else:\n",
+    "        print(\"text\")\n",
+    "        body = {\n",
+    "                    \"inputText\": description,\n",
+    "            }\n",
+    "\n",
+    "    response = bedrock_client.invoke_model(\n",
+    "        body=json.dumps({**body, **embedding_config}), \n",
+    "        modelId=model_id,\n",
+    "        accept=\"application/json\", \n",
+    "        contentType=\"application/json\"\n",
+    "    )\n",
+    "\n",
+    "    return json.loads(response.get(\"body\").read())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "32306437-4931-4450-bc84-999dc8f478a4",
+   "metadata": {},
+   "source": [
+    "Now we can create embeddings for the generated images, together with the product descriptions."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8d3eb696-b5b0-41e5-982b-8da18b9e3978",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "multimodal_embeddings = []\n",
```

Comment:
At this stage, let's:
- only index `image`
- Leave out `description.strip()`
- be explicit on `dimension=xxx`

Reply:
done.